WikiWatershed/model-my-watershed
# Wire In SRAT Catchment API

## Overview

In sub-basin modeling mode, we run the GWLF-E model on each of a HUC-8/HUC-10's composite HUC-12s. The API created at Drexel the "SRAT Catchment API" calculates the attenuates model results for each HUC-12 and its contained NHD Plus catchments.

This PR calls the API with an AOI's HUC-12 results. The API results return to the browser but are not formatted or displayed. In later work we will sum the per HUC-12 attenuated results to get results for the total AOI.

Connects #2612 

### Demo

![screen shot 2018-02-07 at 11 47 10 am](https://user-images.githubusercontent.com/7633670/35936675-ea879442-0c11-11e8-8cec-d1495493c473.png)

[sample_subbasin_response.json.txt](https://github.com/WikiWatershed/model-my-watershed/files/1704409/sample_subbasin_response.json.txt)

### Notes

- [ ] I will add the API key to our fileshare's `default.yml` when this PR has been approved.

## Testing Instructions

 * Pull this branch and provision your app and worker VMs with the api key on Last Pass under "SRAT Catchment API (Staging)" (the worker should be reprovisioned for the security package installation — you can also install these manually with `pip` if you like)
       ```bash
       > MMW_SRAT_CATCHMENT_API_KEY=*** vagrant reload app worker --provision
       ```

<details>
<summary> <strong>NB:</strong> You may wish to give your worker VM two CPUs and run these steps with two celery threads to speed up sub-basin modeling results. This process will involve ...</summary>


```diff
diff --git a/Vagrantfile b/Vagrantfile
index 008818c..1c9dcd4 100644
--- a/Vagrantfile
+++ b/Vagrantfile
@@ -75,6 +75,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |conf
     services.vm.provider "virtualbox" do |v|
       v.memory = 1024
+      v.cpus = 2
     end

     services.vm.provision "ansible" do |ansible|
```

Running `./scripts/debugcelery.sh` and then adding an additional thread:


```sh
> vagrant ssh worker -c \
    "cd /opt/app/ && envdir /etc/mmw.d/env celery -A "mmw.celery:app" worker -l debug -n debug2@%n"
```


</details>
</br>


* Turn the subbasin feature flag on

    ```bash
    > ./scripts/toggle_feature.sh subbasin
    ```
* In the app with your browser's dev tools open, select a HUC-10 and run the MapShed model
* Nothing will be displayed in the sidebar, but the requests should complete and you should see a `results.huc12` object.
